### PR TITLE
Allow disabling of cache cleanup by the end user

### PR DESCRIPTION
**Pull request by ThadHouse**
This behavior is not wanted by all users, and some users actually find it very helpful. This PR adds a property (org.gradle.disable.cache.cleanup) that can be set in the home folder gradle.properties file, and if this property is set to true, the cache deletion will be skipped.

### Context

The caching behavior is useful in many cases, and there are people that don't mind the used space. Having builds always work even in offline scenarios is more important then disk space.

https://github.com/gradle/gradle/issues/6371

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by ThadHouse** [comment]
I would be happy to update the docs too. 

I think this page would need the property
https://docs.gradle.org/current/userguide/build_environment.html

And this page needs to be updated anyway, as the docs saying it only gets downloaded once is not true anymore.
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.wrapper.Wrapper.html

File: subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java [hunk 1]
```diff
@@ -357,10 +357,13 @@ public String getDistributionUrl() {
...
      * <p>If {@link #getGradleVersion()} is not set, will return null.
      *
-     * <p>The wrapper downloads a certain distribution only once and caches it. If your distribution base is the
+     * <p>The wrapper downloads a certain distribution and caches it. If your distribution base is the
      * project, you might submit the distribution to your version control system. That way no download is necessary at
      * all. This might be in particular interesting, if you provide a custom gradle snapshot to the wrapper, because you
      * don't need to provide a download server then.
+     *
+     * After 30 days of no use, the wrapper has the possiblity of being deleted. If this behavior is not desired,
+     * Set the property `org.gradle.disable.cache.cleanup` to true in the gradle.properties file in your gradle home.
```
Comment (kenzierocks) [inline subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java]: nit-pick: I would lower-case `Set` as it isn't the start of a sentence, and will look weird in the generated Javadoc.

File: subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java [hunk 2]
```diff
@@ -27,6 +27,7 @@
 
     private static final long MAX_UNUSED_DAYS_FOR_RELEASES = 30;
     private static final long MAX_UNUSED_DAYS_FOR_SNAPSHOTS = 7;
+    private static final String DISABLE_CACHE_CLEANUP_PROPERTY = "org.gradle.disable.cache.cleanup";
```
Comment (kenzierocks) [inline subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java]: This property doesn't follow what I see as the convention, which is `org.gradle.<item>.<property>`. This should be `org.gradle.cache.cleanup.disabled` in my opinion.

File: subprojects/docs/src/docs/userguide/build_environment.adoc [hunk 3]
```diff
@@ -67,6 +67,9 @@ When set to `all`, `summary` or `none`, Gradle will use different warning type d
 `org.gradle.workers.max=(max # of worker processes)`::
 When configured, Gradle will use a maximum of the given number of workers. Default is number of CPU processors. See also <<command_line_interface.adoc#sec:command_line_performance, performance command-line options>>.
 
+`org.gradle.disable.cache.cleanup=(true,false)`::
+When set to `true`, cache cleanup will be disabled for this gradle home. This property is only settable in the gradle home build.properties file.
```
Comment (kenzierocks) [inline subprojects/docs/src/docs/userguide/build_environment.adoc]: `This property is only settable in the gradle home build.properties file.` reads a little weird. Perhaps ``This property is only read from the `gradle.properties` file in the Gradle home directory.``?

**Comment by ThadHouse** [comment]
I fixed the review issues.

**Comment by ThadHouse** [comment]
Thats a big change to how the caching is implemented, and a complete change in the design. I'm ok doing that, but I would need some help with the gradle internals to figure that one out. How would I change that class to run at the end of builds, and then at that point how would I load properties? Is there some kind of service used?

**Comment by ThadHouse** [comment]
@oehme  Thats a big change to how the caching is implemented, and a complete change in the design. I'm ok doing that, but I would need some help with the gradle internals to figure that one out. How would I change that class to run at the end of builds, and then at that point how would I load properties? Is there some kind of service used?

**Comment by ThadHouse** [comment]
That was kind of my thought. This property makes no sense from a build command perspective, because of you forget it, or don't have it in a specific project but have it in others, it's still useless. So it makes more design sense to me to be only a home property. Maybe it just needs to go in a separate file, not gradle.properties to make that behavior even clearer.

**Comment by oehme** [comment]
It does make sense as a build option the other way round - Disable by default in your user home and then enable for one specific build on the command line for a manual cleanup.

I'd rather avoid another option mechanism.

**Comment by oehme** [comment]
It was not done that way because no one expressed a need to turn it off on the initial thread and everyone wanted Gradle to do this by default. 

I'm on board with adding an option, just not the way it's done here, as it's inconsistent with all other build options. Let's get cache cleanup to the end of the build session, also for the other usability reasons mentioned above. 

Another way would be merging this without the documentation to unblock you and refactor it to the proper solution later and only then announce the option. I'll leave that up to @marcphilipp

**Comment by ThadHouse** [comment]
That works good for me as well. If the undocumented merge gets chosen for now, I'll remove the documentation changes. 

I don't mind if a refactor gets added afterwards, as I am happy to change my setup to accommodate any changes. I would have no problem with a temporary fix changing for a better solution in the future.

**Comment by ThadHouse** [comment]
@marcphilipp I removed the doc changes so this should be good to go in.

File: subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java [hunk 4]
```diff
@@ -20,13 +20,16 @@
...
+import org.gradle.util.GUtil;
 
 import java.io.File;
+import java.util.Properties;
 
 public class GradleUserHomeCleanupService implements Stoppable {
 
     private static final long MAX_UNUSED_DAYS_FOR_RELEASES = 30;
     private static final long MAX_UNUSED_DAYS_FOR_SNAPSHOTS = 7;
+    private static final String DISABLE_CACHE_CLEANUP_PROPERTY = "org.gradle.cache.cleanup.disable";
```
Comment (marcphilipp) [inline subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java]: Please rename this property to `org.gradle.cache.cleanup` and change the logic to that `true` means enabled and `false` disabled.

File: subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java [hunk 5]
```diff
@@ -42,10 +45,18 @@ public GradleUserHomeCleanupService(GradleUserHomeDirProvider userHomeDirProvide
 
     @Override
     public void stop() {
+        File gradleUserHomeDirectory = userHomeDirProvider.getGradleUserHomeDirectory();
+        File gradleProperties = new File(gradleUserHomeDirectory, "gradle.properties");
```
Comment (marcphilipp) [inline subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java]: Please add the following comment above this line:

```java
// TODO Will be implemented without hard-coded access to `$GRADLE_USER_HOME/gradle.properties` for 5.1 in #6084
```

File: subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java (hunk 4 above)
Comment (ThadHouse) [reply subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java]: Should this also mean if the property does not exist that the cache deletion is disabled?

Comment (marcphilipp) [reply subprojects/core/src/main/java/org/gradle/cache/internal/GradleUserHomeCleanupService.java]: No, its default value should be `true`.

**Comment by ThadHouse** [comment]
I fixed the review issue.

**Comment by marcphilipp** [comment]
@oehme PTAL

**Comment by sstock** [comment]
Another use case for this option is Internet connections with a quota.  Downloading a dependency more than once costs money and/or time, depending on the ISP's quota mechanism.  Making very long lived caches important for people in this situation.

**Comment by marcphilipp** [comment]
Thanks, @ThadHouse!

**Comment by ThadHouse** [comment]
Cool! We'll be setting this in our projects Java home and it solves a lot of problems.

**Comment by TaoismDeepLake** [comment]
org.gradle.cache.cleanup
org.gradle.disable.cache.cleanup

which is correct?

**Pull request merged**